Partial duckdb index

lhoestq opened the pull request:
For big datasets, instead of using all the parquet files, use only the first ones until the limit of 5GB is reached.
It's used for both `search` and `filter`.

## Indexing

I added three fields in the duckdb-index job:
- partial (if it used a partial parquet export or if it didn't index the full parquet export)
- num_rows
- num_bytes

## New jobs

I also added two jobs and their responses can be used in the front-end instead of the `size` endpoint:
- config-duckdb-index-size
- dataset-duckdb-index-size

Both jobs return the same response format as the config-size and the dataset-size jobs, but with information about the size of the indexed data instead of the size of the parquet-exported data.

## Update of search and filter

I also added the `partial` field in the responses of `/search` and `/filter`, in case they use a partial index.

## Index file

The file name is `index.duckdb` if it indexes all the rows from the parquet export, otherwise it's named `partial-duckdb.index`. It is placed in the same folder as the parquet export in `refs/convert/parquet`.

Fix https://github.com/huggingface/datasets-server/issues/1742

TODO
- [x] tests
- [x] docs
- [x] openapi
- [x] update /search and /filter for partial indexes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets-server/pr_1992). All of your documentation changes will be reflected on that endpoint.

severo commented:
Reviewing...

About the new jobs: from the frontend point of view, I prefer calling fewer endpoints with more info in them. Would it be OK to move this info to the size cache entries?
An alternative would be to create another endpoint, dedicated to the Hub frontend, that contains all the info we need for a split on page load

wdyt @lhoestq?

severo commented on services/worker/src/worker/config.py:
```diff
@@ -273,9 +273,7 @@ def from_env(cls) -> "DuckDbIndexConfig":
-                max_parquet_size_bytes=env.int(
-                    name="MAX_PARQUET_SIZE_BYTES", default=DUCKDB_INDEX_MAX_PARQUET_SIZE_BYTES
-                ),
+                max_dataset_size=env.int(name="MAX_DATASET_SIZE", default=DUCKDB_INDEX_MAX_DATASET_SIZE),
```
could you update the README?

lhoestq commented:
> About the new jobs: from the frontend point of view, I prefer calling fewer endpoints with more info in them. Would it be OK to move this info to the size cache entries?

I also like the idea of calling fewer endpoints. But in that case shouldn't it call something like `dataset-and-index-size` rather than `size` to get the size of the dataset and index ?

severo commented on libs/libapi/tests/test_duckdb.py:
```diff
@@ -0,0 +1,58 @@
+    split = "split"
+    default_filename = "index.duckdb"
+    split_directory = f"partial-{split}" if partial_split else split
+    filename = f"partial-{default_filename}" if partial_index else default_filename
```
OH ok... So we can have:

`train/index.duckdb`: normal
`train/partial-index.duckdb`: the dataset is originally in Parquet format and contains more than 5GB data: the Parquet "conversion" is complete, while the duckdb index is only on the first 5GB
`partial-train/partial-index.duckdb`: if the Parquet files have been created only on the first 5GB of the data
`partial-train/index.duckdb`: ... we should never have this, right?

lhoestq replied:
`partial-train/index.duckdb` we can have this if the index contains all the data from the partial parquet export

severo replied:
OK. Hence, is it `partial-train/partial-index.duckdb` that will never appear?

severo replied:
well, it should not occur, but... it could, if the parquet files have been created on the first 5GB of data, and the result is N parquet files, where at most N-1 parquet files sum at least 5GB. I don't think it's possible, but... OK

severo commented on services/search/src/search/routes/filter.py:
```diff
@@ -123,15 +123,22 @@ async def filter_endpoint(request: Request) -> Response:
                             revision=revision,
                         )
+
+                    # check if the index is on the full dataset or if it's partial
```
we have the info in `duckdb_index_cache_entry`, no?

lhoestq replied:
not always - I added the `partial` field for this PR and didn't increase the job version

severo replied:
yes, but if I understand correctly, if the duckdb has the `partial` keyword in its name, it also has the field in `duckdb_index_cache_entry`

lhoestq replied:
If it doesn't have the `partial` keyword in its name, the split directory name can still have it have it

severo replied:
ok

severo commented on libs/libapi/src/libapi/response.py:
```diff
@@ -58,3 +62,52 @@ def create_response(
     }
+
+
+def create_maybe_partial_response(
```
wouldn't it make sense to add the partial field to PaginatedResponse?

lhoestq replied:
No because /rows always returns complete data (the rows from 0 to 100 are complete, whereas a search may ignore some rows because it only scans the first 5GB)

severo replied:
We can set `partial: true` for `/rows`. Ideally, all the endpoints that return "rows" would use the same format.

lhoestq replied:
So `/rows` would return `partial: false` all the time.

Ah actually there is one case where it can return `partial: true`: if we request the last page and if we can't show the rest of the dataset.

Let me add `partial` to `/rows` too then

lhoestq replied:
I think I'll just add `partial: true` for all pages of `/rows` if the dataset is partially exported

severo reviewed (commented):
comments on first part of the review

severo commented:
> I also like the idea of calling fewer endpoints. But in that case shouldn't it call something like dataset-and-index-size rather than size to get the size of the dataset and index ?

OK, we can go with these new steps, and then I can create another one that synthesizes everything. It would keep things atomic, and these jobs are ultra-quick and the responses don't occupy much storage.

severo commented on services/worker/src/worker/dtos.py:
```diff
@@ -169,6 +169,63 @@ class ConfigSizeResponse(TypedDict):
+class SplitDuckdbIndex(SplitHubFile):
+    features: Optional[dict[str, Any]]
+    has_fts: bool
+    # The 2 following fields can be None in old cache entries
```
```suggestion
    # The following fields can be None in old cache entries
```

AndreaFrancis commented on services/worker/src/worker/job_runners/config/duckdb_index_size.py:
```diff
@@ -0,0 +1,143 @@
+
+    try:
+        total = 0
+        pending = 0
```
Looks like this var is never used

lhoestq commented:
I took all your comments into account :)

One big change is I added `partial` to `/rows`. It now has `partial: true` if the parquet export for the split is partial

lhoestq commented on services/worker/src/worker/dtos.py:
```diff
@@ -169,6 +169,63 @@ class ConfigSizeResponse(TypedDict):
+class SplitDuckdbIndex(SplitHubFile):
+    features: Optional[dict[str, Any]]
+    has_fts: bool
+    # The following fields can be None in old cache entries
```
I don't think we need an issue or a PR. We should keep the Optional unless we ever need to increase the indexing version.

lhoestq commented:
I factorized the partial checks, added the `bytes` suffix when needed and added a migration to make the `partial` field non-optional (though the `num_rows` and `num_bytes` will still be optional)

severo commented on jobs/mongodb_migration/src/mongodb_migration/migrations/_20231106193200_cache_add_partial_field_in_split_duckdb_index.py:
```diff
@@ -0,0 +1,60 @@
+            },
+            {
+                "$set": {
+                    "content.partial": True,
```
isn't it False?

lhoestq replied:
Good catch ! And actually it should be `None` since it may have indexed a partial parquet dump so we have to keep the `Optional`

severo commented:
and a missing rename in worker
